include/tvm/tir/buffer.h

```diff
@@ -55,8 +55,48 @@ class BufferNode : public Object {
   Var data;
   /*! \brief data type in the content of the tensor */
   DataType dtype;
-  /*! \brief The shape of the buffer */
+  /*! \brief The shape of the buffer
+   *
+   * This contains the shape as it is accessed by
+   * BufferLoad/BufferStore nodes, and used by the low-level code
+   * generators.
+   */
   Array<PrimExpr> shape;
+  /*! \brief The shape of the buffer prior to flattening
+   *
+   * This contains the shape as it exists prior to flattening, and is
+   * used for validating the shape of the tensor passed into the
+   * packed API.
+   *
+   * TODO(Lunderberg): Should this be a reference to the entire
```

Comment:
I'm fine with it either way, though I think I'd want to have that as a later change.  My one concern is that having the entire pre-flattened Buffer available could allow post-flattening passes to have access to more information than is necessary, and information that may be out of date.  Storing only the minimum amount of information required for later use minimizes the amount of state that must be kept consistent during later passes.

Reply:
I've changed my mind, and agree that keeping a reference to the pre-flattened Buffer would be much cleaner.  I needed to add a third similar parameter for the preflattened type, since the post-flattening type of a boolean array is int8, and three similar parameters is too much.

Unfortunately, having a `Optional<Buffer> pre_flattened_buffer` parameter causes some compilation issues.  `BufferNode` cannot be declared before `Buffer`, because `Optional<Buffer>` cannot use a forward declaration.   `Buffer` cannot be declared before `BufferNode`, because `TVM_DEFINE_OBJECT_REF_METHODS` nor the `TVM_DEFINE_OBJECT_REF_COW_METHODS` macros work with a forward declaration.  It is possible to copy-paste declare the methods defined by these macros, then move the definitions themselves into `buffer.cc`, but that has more copy-paste code than I'd like.  I have a rough draft of the changes required, but I think it should be a follow-up PR instead.

Reply:
One additional comment.  The pre-flattened information is feeling less and less like a property of the buffer, and is instead feeling like something that should be stored in the `PrimFuncNode::buffer_map`.  It doesn't represent a difference in the computation, it is solely used to validate user input, and it shouldn't be modified at all after flattening occurs.

Reply:
I like this idea. If this property is only needed for validation, it can be stored in `PrimFuncNode::buffer_map` or maybe `PrimFuncNode::preflattened_buffer_map`

Reply:
It ended up being a relatively straightforward change to add the `PrimFuncNode::preflattened_buffer_map`, which also made it a lot easier to fix some of the breakages caught by `test_lower_build.py`, so that'll be added as another commit on this PR later today.